internal/driver/node.go

```diff
@@ -262,48 +251,32 @@ func (s *nodeServerNoLocked) nodePublishFilesystemVolume(req *csi.NodePublishVol
 	return nil
 }
 
-func (s *nodeServerNoLocked) createDeviceIfNeeded(device string, lv *proto.LogicalVolume) error {
-	var stat unix.Stat_t
-	err := filesystem.Stat(device, &stat)
-	switch err {
-	case nil:
-		// a block device already exists, check its attributes
-		if stat.Rdev == unix.Mkdev(lv.DevMajor, lv.DevMinor) && stat.Uid == uint32(os.Getuid()) && stat.Mode == deviceMode {
-			return nil
-		}
-		err := os.Remove(device)
-		if err != nil {
-			return status.Errorf(codes.Internal, "failed to remove device file %s: error=%v", device, err)
-		}
-		fallthrough
-	case unix.ENOENT:
-		err = os.MkdirAll(path.Dir(device), 0755)
-		if err != nil {
-			return status.Errorf(codes.Internal, "mkdir failed: target=%s, error=%v", path.Dir(device), err)
-		}
-
-		devno := unix.Mkdev(lv.DevMajor, lv.DevMinor)
-		if err := filesystem.Mknod(device, deviceMode, int(devno)); err != nil {
-			return status.Errorf(codes.Internal, "mknod failed for %s. major=%d, minor=%d, error=%v",
-				device, lv.DevMajor, lv.DevMinor, err)
-		}
-	default:
-		return status.Errorf(codes.Internal, "failed to stat %s: error=%v", device, err)
-	}
-	return nil
-}
-
 func (s *nodeServerNoLocked) nodePublishBlockVolume(req *csi.NodePublishVolumeRequest, lv *proto.LogicalVolume) error {
 	// Find lv and create a block device with it
-	targetPath := req.GetTargetPath()
-	err := s.createDeviceIfNeeded(targetPath, lv)
+	mountOption := req.GetVolumeCapability().GetMount()
+	if mountOption == nil {
+		mountOption = &csi.VolumeCapability_MountVolume{}
+	}
+	mountOptions, err := makeMountOptions(req.GetReadonly(), mountOption)
 	if err != nil {
 		return err
 	}
 
+	_, err = os.Create(req.GetTargetPath())
+	if err != nil {
+		return status.Errorf(codes.Internal, "create failed: target=%s, error=%v", req.GetTargetPath(), err)
+	}
+	if err := os.Chmod(req.GetTargetPath(), 0755); err != nil {
+		return status.Errorf(codes.Internal, "chmod failed: target=%s, error=%v", req.GetTargetPath(), err)
+	}
+
+	if err := s.mounter.Mount(lv.GetPath(), req.GetTargetPath(), "", append(mountOptions, "bind")); err != nil {
```

Comment:
Technically this implementation is not a must change, but I think its the only way to implement read only capability for block volumes. Maybe Im missing something though, I will double check.

Reply:
Yeah so before we were mounting a ReadOnly volume as ReadWrite. I know this is not strictly against Kubernetes/CSI spec, but I think if we have the ability to prohibit writes to the volume via bind mount, we should do that.

Reply:
Unlike `volumeMounts`, `volumeDevices` in a Pod object don't have a `readOnly` field. From what I understand, `req.GetReadonly()` should return `true` only when `persistentVolumeClaim.readOnly` in the Pod's `volumes` is set `true`. However, according to [the Kubernetes spec](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.30/#persistentvolumeclaimvolumesource-v1-core) (emphasis mine):

> readOnly Will force the ReadOnly setting in **VolumeMounts**. Default false.

So, it seems like we should ignore this field in `nodePublishBlockVolume`. I'm not completely sure, so please let me know if I'm missing something.

Reply:
I think it is still possible to mark a PVC as readOnly and have that request routed through: 

```yaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: ro-test
  labels:
    type: local
spec:
  storageClassName: lvms-vg1
  resources:
    requests:
      storage: 5Gi
  accessModes:
    - ReadWriteOnce
  volumeMode: Block
---
apiVersion: v1
kind: Pod
metadata:
  name: ro-test
spec:
  volumes:
    - name: data
      persistentVolumeClaim:
        claimName: ro-test
        readOnly: true
  containers:
    - name: container
      image: public.ecr.aws/docker/library/nginx:latest
      ports:
        - containerPort: 80
          name: "http-server"
      volumeDevices:
        - name: data
          devicePath: /dev/xvda
``

This results in the following call

```
{
  "level": "info",
  "ts": "2024-07-11T06:54:21Z",
  "logger": "driver.node",
  "msg": "NodePublishVolume called",
  "volume_id": "88cee7f3-de0a-4a91-a680-6c967a791640",
  "publish_context": null,
  "target_path": "/var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/publish/pvc-2b16260c-a82c-4a7c-af86-4ae8f7c0cffb/42d67280-0d65-4a92-9d17-94c79b7d83f1",
  "volume_capability": "block:<> access_mode:<mode:SINGLE_NODE_WRITER > ",
  "read_only": true,
  "num_secrets": 0,
  "volume_context": {
    "csi.storage.k8s.io/ephemeral": "false",
    "csi.storage.k8s.io/pod.name": "ro-test",
    "csi.storage.k8s.io/pod.namespace": "openshift-storage",
    "csi.storage.k8s.io/pod.uid": "42d67280-0d65-4a92-9d17-94c79b7d83f1",
    "csi.storage.k8s.io/serviceAccount.name": "default",
    "storage.kubernetes.io/csiProvisionerIdentity": "1720678205542-7248-topolvm.io"
  }
}
{
  "level": "info",
  "ts": "2024-07-11T06:54:21Z",
  "logger": "driver.node",
  "msg": "NodePublishVolume(block) succeeded",
  "volume_id": "88cee7f3-de0a-4a91-a680-6c967a791640",
  "target_path": "/var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/publish/pvc-2b16260c-a82c-4a7c-af86-4ae8f7c0cffb/42d67280-0d65-4a92-9d17-94c79b7d83f1"
}
```

I agree that the k8s filesystem protection maybe enough, but I think a bind mount is safer and also recommended because then we can later introudce Stage/UnstageVolume without problems

Reply:
Thanks for your detailed explanation. I feel that I now have a much better understanding of your points.

Although this is a (kind of) bug fix, it does change the current behavior of TopoLVM. Could you separate this part into a new PR? This would make the release process much easier.

Reply:
Can you tell me what you had in mind as replacement for https://github.com/topolvm/topolvm/pull/930/files#diff-1390f622511e6fe8aa2d84381b2cb30c1126bce425b8d8f404b758d03b527f7fR271 ? I would need to understand how to change the code without relying on /dev/topolvm

Reply:
As you mentioned, we need to keep using mknod(2), and `req.GetTargetPath()` might point to a different path than /dev/vg/lv. However, the target path won't be /dev/topolvm/*, so we won't call mknod(2) for /dev/topolvm in `nodePublishBlockVolume`. Therefore, we can stop using /dev/topolvm entirely.

> Can you tell me what you had in mind as replacement for https://github.com/topolvm/topolvm/pull/930/files#diff-1390f622511e6fe8aa2d84381b2cb30c1126bce425b8d8f404b758d03b527f7fR271 ?

What I meant is that we should call mknod(2) in the same way as before in `nodePublishBlockVolume`. Here's the code:
```go
func (s *nodeServerNoLocked) nodePublishBlockVolume(req *csi.NodePublishVolumeRequest, lv *proto.LogicalVolume) error {
	// Find lv and create a block device with it
	targetPath := req.GetTargetPath()
	err := s.createDeviceIfNeeded(targetPath, lv) // this calls mknod(2).
	if err != nil {
		return err
	}
	// ...
}
```

Reply:
The point of the PR was to drop mknod from topolvm. If we keep this then this is effectively a rename from the path that lvm2 reports. I do not think we should adopt this.

Reply:
I believe this is regression free and overall should be the way to go, since kubelet upstream also handles readonly mounts via bind. I do not believe this to be problematic

Reply:
Thanks for your explanation. After talking it over with my colleagues, I've changed my mind. I agree with you that we don't need to modify this part.